discovery/manager.go

```diff
@@ -153,41 +146,41 @@ func (m *Manager) cancelDiscoverers() {
 	for _, c := range m.discoverCancel {
 		c()
 	}
-	m.targets = make(map[poolKey][]*targetgroup.Group)
+	m.targets = make(map[poolKey]map[string]*targetgroup.Group)
 	m.discoverCancel = nil
 }
 
-func (m *Manager) addGroup(poolKey poolKey, tg []*targetgroup.Group) {
+func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
-		if tg != nil {
-			m.targets[poolKey] = tg
+		if tgs == nil {
+			close(done)
+			return
+		}
+
+		for _, tg := range tgs {
+			if tg != nil { // Some Discoverers send nil targetgroup so need to check for it to avoid panics.
+				if _, ok := m.targets[poolKey]; !ok {
+					m.targets[poolKey] = make(map[string]*targetgroup.Group)
+				}
+				m.targets[poolKey][tg.Source] = tg
```

Comment:
Check if `len(tg.Targets) == 0`. This means those targets went away and we can remove the group from the map. So if a SD has a target per source, and is highly dynamic, the map simply grows and we might have a memory leak.

Reply:
In hindsight, this would break the discovery further down where we call `allGroups()` and we might not send anything against a `setName` and the old targets still hang around. 

Should we worry about the memory leak though?

Reply:
I was worried that the map will grow indefinitely   as well, but this is how it was implemented before so I guess it wasn't a problem. Anyway I added some  code to delete empty maps